### DISCUSSION TITLE
How to read the defined schema rules

### DISCUSSION THREAD

**Discussion by bartosjiri** [colinhacks/zod#1857] [Q&A] [answered]
How can I read the values and other definitions of a Zod schema?

**Example:**
Having the following schema:
```ts
const exampleSchema = z.object({
  title: z.string().min(3),
  description: z.string().optional().min(5)
})
```
How can I use this schema in the frontend application, apart from validation, to also provide constrains for my form. I would like the schema to be the single source of truth and be able to mark my `title` form field as required and the `description` text input field to have the native `minlength` set to 5, using the values from this schema.

**Comment by bartosjiri** [comment]
Think I figured it out, in case anyone else is looking for a similar solution:

To read the `optional` value:
```ts
schema['<field-name>'].isOptional()
```

To read any limit value, e.g. `min`:
```ts
schema['<field-name>']._def.checks?.find(({kind}) => kind === 'min')?.value
```

**Reply by JacobWeisenburger** [reply]
this is almost correct. You need to use `.shape` to access the field schemas.

**Reply by bartosjiri** [reply]
Indeed, I destructured earlier and copy-pasted incorrectly only the rest of the code in here. Thanks!

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const exampleSchema = z.object( {
    title: z.string().min( 3 ),
    description: z.string().min( 5 ).optional()
} )

console.log( exampleSchema.shape.description.isOptional() ) // true

const minCheck = exampleSchema.shape.description
    ._def.innerType
    ._def.checks?.find( ( { kind } ) => kind === 'min' )
if ( minCheck?.kind == 'min' ) console.log( minCheck.value ) // 5
```

**Reply by bartosjiri** [reply]
Maybe for the completeness of the solution it is worth mentioning that the depth level of the `checks` property is dependent on the rule chaining. If implementing, I suggest using a recursive lookup based on presence of the `innerType`/`checks` properties. Please correct me if this is wrong.
